core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/AsyncResponseHandler.java

```diff
@@ -73,6 +73,9 @@ public void onStream(Publisher<ByteBuffer> publisher) {
 
     @Override
     public void onError(Throwable err) {
+        if (streamFuture == null) {
+            prepare();
+        }
```

Comment:
For my understanding:
* so onError can possibly get called without prepare()?
* calling prepare here and then immediately calling completeExceptionally on it. is the idea to just make sure streamFuture is non-null or is also setting up the thenComplete important? Wondering about just doing `streamFuture = new CompletableFuture<>();` here instead of `prepare()`, and further why not initialize `streamFuture = new CompletableFuture<>()` in the field declaration itself instead of in `prepare()`.

Reply:
FWIU, `prepare()` would be called in the normal flow to get the transformed response (see `MakeAsyncHttpRequestStage`), so if there is an error after `prepare()` is called there, then `streamFuture` wouldn't be null. On the other hand, if there is an error before the normal flow is executed (e.g. `prepare()` wasn't called before an error occured), then `prepare` is called here so that the response-handler is called before exceptionally-completing the stream-future.

Reply:
> so onError can possibly get called without prepare()?

Yes. We aren't clear in the docs, but it can happen as it happened in this case. I think in general our use of prepare() should be pushed upstream more, since we have other cases where prepare() isn't getting called soon enough. I wanted to have a light touch here, since I know this is a fragile part of the codebase.

Reply:
If we did `streamFuture = new CompletableFuture<>()` instead of `prepare()`, the behavior would be the same, assuming `prepare()` is defined as it is today. Which we should use depends on what we think is safer in the future.

Initializing streamFuture in the field declaration is riskier. We don't really document the lifecycle of a handler, so if the same handler is used for retries it wouldn't work. I don't know exactly which we do.

Reply:
Thanks Matt for helping me understand better. At least my questions seem ballpark relevant. Agree with the light touch approach - this change seems safest option to me right now.